File: solutions/go/minesweeper/1/minesweeper.go
package minesweeper

func Annotate(board []string) []string {
    annotated := []string{}
	for i, row := range board {
        new_row := ""
        for j, ch := range row {
            if ch == ' ' {
                count := count_bombs(i, j, board)
                new_row += digit_to_char(count)
            } else {
                new_row += "*"
            }
        }
        annotated = append(annotated, new_row)
    }
    return annotated
}

func digit_to_char(count int) string {
    switch(count) {
    case 1:
    	return "1"
    case 2:
    	return "2" 
	case 3:
    	return "3" 
    case 4:
    	return "4" 
    case 5:
    	return "5" 
    case 6:
    	return "6" 
    case 7:
    	return "7" 
    case 8:
    	return "8" 
    default:
    	return " "  
    }
}

func count_bombs(x, y int, board []string) (count int) {
    for i := -1; i < 2; i++ {
        for j := -1; j < 2; j++ {
            if i == 0 && j == 0 {
            	continue
            }
			if x + i >= 0 &&  x + i < len(board) && y + j >= 0 && y + j < len(board[x + i]) {
           		if board[x + i][y + j] == '*' {
                    count++
                }         
            }
            
        }
    }
    return count
}

File: solutions/clojure/minesweeper/1/src/minesweeper.clj
(ns minesweeper
  (:require [clojure.string :as s]))

(defn make-position-validator [rows-num columns-num]
  (fn [x y]
    (and (< -1 x rows-num)
         (< -1 y columns-num))))

(defn neighbors [[x y] valid-f]
  (for [[update-x update-y] [[inc inc] [dec inc]
                             [dec dec] [inc dec]
                             [inc identity] [dec identity]
                             [identity inc] [identity dec]]
        :let [neighbor-x (update-x x)
              neighbor-y (update-y y)]
        :when (valid-f neighbor-x neighbor-y)]
    [neighbor-x neighbor-y]))

(defn string->board [s]
  (->> s
       (s/split-lines)
       (mapv #(s/split % #""))))

(defn bomb? [cell] (= "*" cell))

(defn draw-cell? [cell value]
  (and (= " " cell)
       (pos? value)))

(defn draw [input]
  (let [board (string->board input)
        rows (count board)
        columns (count (nth board 0))
        final-board (for [x (range 0 rows) y (range 0 columns)
                          :let [valid? (make-position-validator rows columns)
                                neighbors (neighbors [x y] valid?)
                                cell (get-in board [x y])
                                value (count (filter bomb? (map #(get-in board %) neighbors)))]]
                      (if (draw-cell? cell value) value cell))]
    (s/join "\n" (map (partial apply str) (partition columns final-board)))))
